go-controller/pkg/factory/factory_test.go

```diff
@@ -558,7 +576,9 @@ var _ = Describe("Watch Factory Operations", func() {
 			for _, ot := range testNodes {
 				nodes = append(nodes, ot.node)
 				nodeWatch.Add(ot.node)
+				ot.mu.Lock()
 				ot.node.Status.Phase = v1.NodeTerminated
```

Comment:
OK, so the race checker appears to be complaining that we write to this variable here after having read it previously in the `UpdateFunc` above. That is, it is complaining that the _second_ time we do `ot.node.Status.Phase = v1.NodeTerminated` might happen before, after, or during the _first_ time we run `UpdateFunc`.

That is, in the main goroutine we do
```
1. nodeWatch.Add
2. ot.node.Status.Phase = v1.NodeTerminated
3. nodeWatch.Modify
4. nodeWatch.Delete
5. nodeWatch.Add
6. ot.node.Status.Phase = v1.NodeTerminated
7. nodeWatch.Modify
8. nodeWatch.Delete
```

and step 3 guarantees that we eventually call `UpdateFunc`, but doesn't guarantee whether we call it before or after step 6. As it happens, it doesn't matter whether we call it before or after step 6, since step 6 is a no-op. The only actual problem would occur if we called `UpdateFunc` _during_ step 6, and adding a mutex ensures that that can't happen. 

Honestly, it's not clear what the point of this code is anyway; we never test that `Phase` _isn't_ set to `NodeTerminated` anywhere, so we could just move this assignment to before the loop and get the same result...

Reply:
IIRC the point was just to have *something* to update on Node. Not that the value mattered.

Reply:
But it's not testing anything useful. It's just ensuring that `UpdateFunc()` does not get called before the first time we call `nodeWatch.Modify()`, and _nothing else_. It's not like we're testing that updates don't get queued when nothing has changed, or something like that. We could just as easily be using an unrelated global variable here.